SSH2Stream: Dispatch overloaded USERAUTH packet types (60, 61)

andrewjjenkins opened the pull request:
The SSH protocol interprets packets with type 60 differently based on
the authentication method currently in progress (see RFC4252 and the
keyboard-interactive internet-draft):
  publickey:  USERAUTH_PK_OK
  keyboard-interactive: USERAUTH_INFO_REQUEST  (prompt)
  password: USERAUTH_PASSWORD_CHANGEREQ

Thus ssh2-streams needs to keep track of what authentication is in
progress when it is parsing a packet.  Whenever a new authentication
method is attempted, that's the one that is in progress until some sort
of success or failure is received from the server (USERAUTH_SUCCESS,
USERAUTH_FAILURE, or USERAUTH_PK_OK).  Keep the authentication methods
in progress in a queue.

Add test/test-packet60 with some basic packet tests and a client that
blindly sends a public key check, public key auth, password auth, and
keyboard interactive auth at once, exercising the client/server auth
queue.

mscdex commented on lib/ssh.js:
```diff
@@ -4011,11 +4035,16 @@ function parse_USERAUTH(self, type, callback) {
+  } else if (authMethod === undefined) {
+    self.emit('error', new Error('Packet 60 but no auth in progress'));
+  } else {
+    self.emit(new Error('Packet 60 invalid for ' + authMethod));
```
Missing `'error'` as first argument here.

mscdex commented on lib/ssh.js:
```diff
@@ -1679,7 +1691,7 @@ SSH2Stream.prototype.authNone = function(username) {
   buf.write('none', p += 4, 4, 'ascii');
 
-  this._state.authMethod = 'none';
+  instate.authsQueue.push('keyboard-interactive');
```
This should be `'none'`.

mscdex commented on lib/ssh.js:
```diff
@@ -613,6 +617,7 @@ SSH2Stream.prototype.reset = function(noend) {
         kexinit: undefined,
         identRaw: undefined,
         rekeyQueue: [],
+        authsQueue: [],
```
Since we're actually using `authsQueue` for both inbound for servers and outbound for clients, maybe we move it up a level so it's directly on the `_state` object. Also `authMethod` should be removed since the queue supersedes it.

andrewjjenkins commented:
Thanks for the review.  I think I addressed all your suggestions.  Once we're happy, I can squash into one commit.

andrewjjenkins commented:
I collapsed into one commit.  I believe it addresses all your review comments.  @mscdex please let me know if you see anything else wrong/needing improvement.

mscdex commented on test/test-packet60.js:
```diff
@@ -0,0 +1,267 @@
+}
+
+function signWithClientKey(blob, syncCb) {
+  var signType = (PARSED_CLIENT_PRV_KEY.type ===
```
Style nit: a ternary expression should be broken on `?` and `:` like:

```javascript
(PARSED_CLIENT_PRV_KEY.type === 'rsa'
 ? 'RSA-SHA1'
 : 'DSA-SHA1');
```

mscdex commented on test/test-packet60.js:
```diff
@@ -0,0 +1,267 @@
+    assert.equal(method, 'keyboard-interactive');
+    assert.equal(data, '');
+    process.nextTick(function () {
+      server.authInfoReq('req 0', 'instructions',
```
Style nit: try to keep arguments either all on the same line or split all of them (indented and lined up) for consistency, like:

```javascript
server.authInfoReq('req 0',
                   'instructions',
                   [ { prompt: 'Say something to req 0', echo: true } ]);
```

mscdex commented on test/test-packet60.js:
```diff
@@ -0,0 +1,267 @@
+      assert.equal(responses.length, 1);
+      assert.equal(responses[0], 'hello to req 0');
+      process.nextTick(function () {
+        server.authInfoReq('req 1', 'new instructions',
```
Same here

mscdex commented on test/test-packet60.js:
```diff
@@ -0,0 +1,267 @@
+      assert.equal(name, 'req 0');
+      assert.equal(inst, 'instructions');
+      assert.equal(lang, '');
+      assert.deepEqual(prompts,
```
Same here

mscdex commented on lib/ssh.js:
```diff
@@ -1951,6 +1964,7 @@ SSH2Stream.prototype.authPKOK = function(keyAlgo, key) {
   buf.writeUInt32BE(keyLen, 5 + keyAlgoLen, true);
   key.copy(buf, 5 + keyAlgoLen + 4);
 
+  this._state.authsQueue.shift();
```
Perhaps for method-specific responses like this one, we should verify that the current method in the queue is the right/expected type (e.g. `publickey` in this particular case).

mscdex commented:
A few issues, but otherwise LGTM.

mscdex commented:
Landed in 7831f2499c2f9a8ee25e1aea25cd4d77ba848d21 with a few changes and updates. Thanks!
